File: core/src/main/java/com/cartenz/core/api/MySubscriber.java
package com.cartenz.core.api;

import android.text.TextUtils;
import android.util.Log;

import com.cartenz.core.utils.Dictionary;
import com.google.gson.Gson;

import java.util.ArrayList;
import java.util.List;

import rx.Subscriber;

public abstract class MySubscriber<T> extends Subscriber<T> {

    @Override
    public void onCompleted() {
        onComplete();
    }

    @Override
    public void onError(Throwable e) {
        Log.wtf("Test_e", e + "");
        onError(e.getMessage(), "");
        onComplete();
    }

    @Override
    public void onNext(T t) {
        Gson gson = new Gson();
        String json = gson.toJson(t);
        BaseApiDao baseApiDao = gson.fromJson(json, BaseApiDao.class);
        String message = null;
        Log.wtf("test_okhttp_yy", t + " " + json + " " + baseApiDao.message + " " + baseApiDao.errors);

        if (!TextUtils.isEmpty(baseApiDao.message)) {
            message = baseApiDao.message;
        }

        String errorMessage = null;
        if (baseApiDao.errors != null) {
            List<BaseApiDao.Errors> baseErrorDaoList = baseApiDao.errors;
            List<String> error = new ArrayList<>();
            for (int i = 0; i < baseErrorDaoList.size(); i++) {
                Log.wtf("Test_" + i, baseErrorDaoList.get(i).itemCode + "");
                error.add(baseErrorDaoList.get(i).itemCode + "");
            }
            Log.wtf("test_e", error + " " + error.size());
            errorMessage = Dictionary.getValueByKey(error);
        }

        Log.wtf("test_okhttp_x", message + " " + errorMessage);

        if (TextUtils.isEmpty(message) && TextUtils.isEmpty(errorMessage)) {
            onSuccess(t);
        } else {
            if (TextUtils.isEmpty(message)) {
                message = "";
            }
            if (TextUtils.isEmpty(errorMessage)) {
                errorMessage = "";
            }
            onError(message, errorMessage);
        }

        onComplete();
    }

    public abstract void onError(String message, String ErrorMessage);

    public abstract void onSuccess(T t);

    public abstract void onComplete();
}

File: cartenz/src/main/java/com/cartenz/cartenz/api/MySubscriber.kt
package com.cartenz.kotlin_core.api

import android.text.TextUtils
import com.cartenz.cartenz.Dictionary
import com.google.gson.Gson
import rx.Subscriber
import java.util.*

abstract class MySubscriber<T> : Subscriber<T>() {

    override fun onCompleted() {
        onComplete()
    }

    override fun onError(e: Throwable) {
        onError(e.message!!, "")
        onComplete()
    }

    override fun onNext(t: T) {
        val gson = Gson()
        val json = gson.toJson(t)
        val baseApiDao = gson.fromJson(json, BaseApiDao::class.java)
        var message: String? = null

        if (!TextUtils.isEmpty(baseApiDao.message)) {
            message = baseApiDao.message
        }

        var errorMessage: String? = null
        if (baseApiDao.errors != null) {
            val baseErrorDaoList = baseApiDao.errors
            val error = ArrayList<String>()
            for (i in baseErrorDaoList?.indices!!) {
                error.add(baseErrorDaoList?.get(i).itemCode.toString() + "")
            }
            errorMessage = Dictionary.getValueByKey(error)
        }
        if (!TextUtils.isEmpty(message) || !TextUtils.isEmpty(errorMessage) || baseApiDao.data == null || baseApiDao.errors != null && baseApiDao.errors!!.size > 0) {
            if (TextUtils.isEmpty(message)) {
                message = ""
            }
            if (TextUtils.isEmpty(errorMessage)) {
                errorMessage = ""
            }
            onError(message, errorMessage)
        } else {
            onSuccess(t)
        }

        onComplete()
    }

    abstract fun onError(message: String?, errorMessage: String?)

    abstract fun onSuccess(t: T)

    abstract fun onComplete()


}